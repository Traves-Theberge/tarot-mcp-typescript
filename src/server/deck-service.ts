/**
 * Deck service for drawing and managing tarot cards
 */

import { TarotCard, TarotCardError, createFullDeck, DECK_CONSTANTS } from '../core/index.js';
import {
  RandomNumberGenerator,
  SystemRandomNumberGenerator,
  shuffleArray,
} from '../utils/index.js';

/**
 * Service class for managing tarot deck operations
 */
export class TarotDeckService {
  private static readonly FULL_DECK = createFullDeck();
  private static readonly SHUFFLE_ITERATIONS = 10;

  constructor(private readonly rng: RandomNumberGenerator = new SystemRandomNumberGenerator()) {}

  /**
   * Draw cards from the deck
   * 
   * @param count - Number of cards to draw (1-78)
   * @returns Array of drawn cards
   * @throws TarotCardError if count is invalid
   */
  drawCards(count: number): readonly TarotCard[] {
    this.validateCardCount(count);

    // Create a copy of the full deck
    let deck = [...TarotDeckService.FULL_DECK];

    // Shuffle the deck multiple times for better randomness
    for (let i = 0; i < TarotDeckService.SHUFFLE_ITERATIONS; i++) {
      deck = shuffleArray(deck, this.rng);
    }

    // Return the requested number of cards
    return deck.slice(0, count);
  }

  /**
   * Get the complete tarot deck (78 cards)
   * 
   * @returns Array of all tarot cards
   */
  getFullDeck(): readonly TarotCard[] {
    return [...TarotDeckService.FULL_DECK];
  }

  /**
   * Get the total number of cards in a complete deck
   */
  get deckSize(): number {
    return DECK_CONSTANTS.TOTAL_CARDS;
  }

  /**
   * Validate card count is within valid range
   */
  private validateCardCount(count: number): void {
    if (count < 1 || count > DECK_CONSTANTS.TOTAL_CARDS) {
      throw new TarotCardError(`Invalid card count: ${count}. Must be between 1 and ${DECK_CONSTANTS.TOTAL_CARDS}.`);
    }
  }
} 