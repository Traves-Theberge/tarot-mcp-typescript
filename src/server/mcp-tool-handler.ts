/**
 * MCP tool handler for tarot server operations
 */

import {
  Tool,
  CallToolResult,
  ImageContent,
} from '@modelcontextprotocol/sdk/types.js';

import { TarotCard, CardResponse, getCardName } from '../core/index.js';
import { generateCardURI, parseCardURI, loadCardImageAsBase64, ImageLoadError } from '../utils/index.js';
import { TarotDeckService } from './deck-service.js';

/**
 * Request types for MCP tools
 */
interface DrawCardsRequest {
  count?: number;
}

interface FetchImagesRequest {
  uris: string[];
}

/**
 * MCP tool handler for tarot functionality
 */
export class TarotMCPToolHandler {
  constructor(private readonly deckService: TarotDeckService) {}

  /**
   * Get list of available MCP tools
   */
  getAvailableTools(): Tool[] {
    return [
      {
        name: 'draw_cards',
        description: 'Draw one or more tarot cards',
        inputSchema: {
          type: 'object',
          properties: {
            count: {
              type: 'integer',
              description: 'Number of cards to draw (1-78)',
              minimum: 1,
              maximum: 78,
              default: 1,
            },
          },
        },
      },
      {
        name: 'get_full_deck',
        description: 'Get all 78 tarot cards in the deck',
        inputSchema: {
          type: 'object',
          properties: {},
        },
      },
      {
        name: 'fetch_images',
        description: 'Fetch base64-encoded images for the given tarot card URIs',
        inputSchema: {
          type: 'object',
          properties: {
            uris: {
              type: 'array',
              description: 'Array of tarot card URIs to fetch images for',
              items: {
                type: 'string',
                pattern: '^tarot://card/(major/\\d+|minor/(wands|pentacles|swords|cups)/\\d+)$',
              },
            },
          },
          required: ['uris'],
        },
      },
    ];
  }

  /**
   * Handle MCP tool calls
   */
  async handleToolCall(name: string, args?: unknown): Promise<CallToolResult> {
    try {
      switch (name) {
        case 'draw_cards':
          return this.handleDrawCards(args as DrawCardsRequest);
        case 'get_full_deck':
          return this.handleGetFullDeck();
        case 'fetch_images':
          return this.handleFetchImages(args as FetchImagesRequest);
        default:
          throw new Error(`Unknown tool: ${name}`);
      }
    } catch (error) {
      return {
        content: [
          {
            type: 'text',
            text: `Error: ${error instanceof Error ? error.message : String(error)}`,
          },
        ],
        isError: true,
      };
    }
  }

  /**
   * Handle draw_cards tool
   */
  private handleDrawCards(args?: DrawCardsRequest): CallToolResult {
    const count = args?.count ?? 1;
    const cards = this.deckService.drawCards(count);
    return this.createCardsResponse(cards);
  }

  /**
   * Handle get_full_deck tool
   */
  private handleGetFullDeck(): CallToolResult {
    const cards = this.deckService.getFullDeck();
    return this.createCardsResponse(cards);
  }

  /**
   * Handle fetch_images tool
   */
  private handleFetchImages(args: FetchImagesRequest): CallToolResult {
    if (!args?.uris || !Array.isArray(args.uris)) {
      throw new Error('uris parameter is required and must be an array');
    }

    const content: ImageContent[] = [];

    for (const uri of args.uris) {
      try {
        const card = parseCardURI(uri);
        if (!card) {
          throw new Error(`Invalid tarot card URI: ${uri}`);
        }

        const base64Data = loadCardImageAsBase64(card);
        content.push({
          type: 'image',
          data: base64Data,
          mimeType: 'image/png',
        });
      } catch (error) {
        if (error instanceof ImageLoadError) {
          throw error;
        }
        throw new Error(`Failed to load image for URI: ${uri}`);
      }
    }

    return { content };
  }

  /**
   * Create a response with card data in JSON format
   */
  private createCardsResponse(cards: readonly TarotCard[]): CallToolResult {
    const cardResponses: CardResponse[] = cards.map(card => ({
      name: getCardName(card),
      uri: generateCardURI(card),
    }));

    const jsonString = JSON.stringify(cardResponses, null, 2);

    return {
      content: [
        {
          type: 'text',
          text: jsonString,
        },
      ],
    };
  }
} 