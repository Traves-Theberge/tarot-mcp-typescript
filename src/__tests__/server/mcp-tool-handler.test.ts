/**
 * Tests for MCP Tool Handler functionality
 * TypeScript equivalent of Swift TarotServerHandlerTests
 */

import { TarotCardError } from '../../core/index.js';
import { TarotMCPToolHandler } from '../../server/mcp-tool-handler.js';
import { TarotDeckService } from '../../server/deck-service.js';
import { SeedablePseudoRNG } from '../test-support/index.js';

describe('MCP Tool Handler Tests', () => {
  describe('TarotToolRequest Error Handling', () => {
    test('init error description is good', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(42));
      const handler = new TarotMCPToolHandler(deckService);
      
      const result = await handler.handleToolCall('draw_cards', { count: 0 });
      expect(result.isError).toBe(true);
      expect(result.content[0]?.type).toBe('text');
      if (result.content[0]?.type === 'text') {
        expect(result.content[0].text).toContain('Invalid card count: 0. Count must be between 1 and 78.');
      }
    });
  });

  describe('draw_cards tool', () => {
    test('returns valid response for single card', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(42));
      const handler = new TarotMCPToolHandler(deckService);
      const result = await handler.handleToolCall('draw_cards', {});

      expect(result.content).toHaveLength(1);
      expect(result.content[0]?.type).toBe('text');
      
      if (result.content[0]?.type === 'text') {
        const jsonData = JSON.parse(result.content[0].text as string);
        expect(Array.isArray(jsonData)).toBe(true);
        expect(jsonData).toHaveLength(1);
        
        const card = jsonData[0];
        expect(typeof card.name).toBe('string');
        expect(typeof card.imageURI).toBe('string');
      }
    });

    test('produces deterministic results for single card', async () => {
      const deckService1 = new TarotDeckService(new SeedablePseudoRNG(54321));
      const deckService2 = new TarotDeckService(new SeedablePseudoRNG(54321));
      const handler1 = new TarotMCPToolHandler(deckService1);
      const handler2 = new TarotMCPToolHandler(deckService2);

      const result1 = await handler1.handleToolCall('draw_cards', {});
      const result2 = await handler2.handleToolCall('draw_cards', {});

      expect(result1.content).toHaveLength(1);
      expect(result2.content).toHaveLength(1);

      if (result1.content[0]?.type === 'text' && result2.content[0]?.type === 'text') {
        expect(result1.content[0].text).toBe(result2.content[0].text);
        
        // Snapshot test for deterministic output
        expect(result1.content[0].text).toMatchSnapshot();
      }
    });

    test('with default count', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(98765));
      const handler = new TarotMCPToolHandler(deckService);
      const result = await handler.handleToolCall('draw_cards', {});

      expect(result.content).toHaveLength(1);
      
      if (result.content[0]?.type === 'text') {
        const jsonData = JSON.parse(result.content[0].text as string);
        expect(jsonData).toHaveLength(1);
        
        const card = jsonData[0];
        expect(card.name).toBe('Ten of Cups');
        expect(card.imageURI).toBe('tarot://card/minor/cups/10');
      }
    });

    test('with custom count', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(11111));
      const handler = new TarotMCPToolHandler(deckService);
      const result = await handler.handleToolCall('draw_cards', { count: 7 });

      expect(result.content).toHaveLength(1);
      
      if (result.content[0]?.type === 'text') {
        const jsonData = JSON.parse(result.content[0].text as string);
        expect(jsonData).toHaveLength(7);

        // Verify each card has name and image
        for (const card of jsonData) {
          expect(typeof card.name).toBe('string');
          expect(typeof card.imageURI).toBe('string');
          expect(card.imageURI).toMatch(/^tarot:\/\/card\//);
        }
      }
    });

    test('returns errors for invalid counts', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(22222));
      const handler = new TarotMCPToolHandler(deckService);

      // Test upper bound validation
      const result1 = await handler.handleToolCall('draw_cards', { count: 79 });
      expect(result1.isError).toBe(true);

      // Test lower bound validation
      const result2 = await handler.handleToolCall('draw_cards', { count: 0 });
      expect(result2.isError).toBe(true);

      // Test negative values
      const result3 = await handler.handleToolCall('draw_cards', { count: -5 });
      expect(result3.isError).toBe(true);
    });
  });

  describe('get_full_deck tool', () => {
    test('returns complete deck', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(42));
      const handler = new TarotMCPToolHandler(deckService);
      const result = await handler.handleToolCall('get_full_deck', {});

      expect(result.content).toHaveLength(1);
      
      if (result.content[0]?.type === 'text') {
        const jsonData = JSON.parse(result.content[0].text as string);
        expect(jsonData).toHaveLength(78);

        // Check that specific cards are present
        const cardNames = jsonData.map((card: any) => card.name);
        expect(cardNames).toContain('The Fool');
        expect(cardNames).toContain('Ace of Wands');
        expect(cardNames).toContain('King of Cups');

        // Verify each card has name and image
        for (const card of jsonData) {
          expect(typeof card.name).toBe('string');
          expect(typeof card.imageURI).toBe('string');
          expect(card.imageURI).toMatch(/^tarot:\/\/card\//);
        }
      }
    });

    test('produces deterministic results', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(33333));
      const handler = new TarotMCPToolHandler(deckService);
      const result = await handler.handleToolCall('get_full_deck', {});

      expect(result.content).toHaveLength(1);
      
      if (result.content[0]?.type === 'text') {
        // Snapshot test for deterministic output
        expect(result.content[0].text).toMatchSnapshot();
      }
    });
  });

  describe('fetch_images tool', () => {
    test.skip('returns images for valid URIs', async () => {
      // Skipped because image files are not available in test environment
      const deckService = new TarotDeckService(new SeedablePseudoRNG(42));
      const handler = new TarotMCPToolHandler(deckService);
      const uris = [
        'tarot://card/major/0',
        'tarot://card/minor/cups/1'
      ];
      
      const result = await handler.handleToolCall('fetch_images', { uris });

      expect(result.content).toHaveLength(2);
      
      for (const content of result.content) {
        expect(content.type).toBe('image');
        if (content.type === 'image') {
          expect(typeof content.data).toBe('string');
          expect(content.mimeType).toBe('image/png');
        }
      }
    });

    test('returns error for invalid URIs', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(42));
      const handler = new TarotMCPToolHandler(deckService);
      const invalidUris = ['invalid://uri'];
      
      const result = await handler.handleToolCall('fetch_images', { uris: invalidUris });
      expect(result.isError).toBe(true);
    });
  });

  describe('Error Handling', () => {
    test('unknown tool returns error', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(42));
      const handler = new TarotMCPToolHandler(deckService);

      const result = await handler.handleToolCall('unknown_tool', {});
      expect(result.isError).toBe(true);
    });

    test('error propagation works correctly', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(42));
      const handler = new TarotMCPToolHandler(deckService);

      // Test that validation errors are properly propagated
      const result1 = await handler.handleToolCall('draw_cards', { count: 0 });
      expect(result1.isError).toBe(true);

      const result2 = await handler.handleToolCall('draw_cards', { count: 79 });
      expect(result2.isError).toBe(true);
    });
  });

  describe('Tool Configuration', () => {
    test('getAvailableTools returns correct tool list', () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(42));
      const handler = new TarotMCPToolHandler(deckService);
      const tools = handler.getAvailableTools();

      expect(tools).toHaveLength(3);

      const toolNames = tools.map(tool => tool.name);
      expect(toolNames).toContain('draw_cards');
      expect(toolNames).toContain('get_full_deck');
      expect(toolNames).toContain('fetch_images');

      // Verify draw_cards has proper schema
      const drawCardsTool = tools.find(tool => tool.name === 'draw_cards');
      expect(drawCardsTool).toBeDefined();
      expect(drawCardsTool?.inputSchema).toBeDefined();
    });

    test('tools are configured correctly', () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(42));
      const handler = new TarotMCPToolHandler(deckService);
      const tools = handler.getAvailableTools();
      
      expect(tools).toHaveLength(3);
      expect(tools.some(tool => tool.name === 'draw_cards')).toBe(true);
      expect(tools.some(tool => tool.name === 'get_full_deck')).toBe(true);
      expect(tools.some(tool => tool.name === 'fetch_images')).toBe(true);
    });

    test('handlers respond correctly', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(12345));
      const handler = new TarotMCPToolHandler(deckService);

      // Test equivalent to ListTools
      const tools = handler.getAvailableTools();
      expect(tools).toHaveLength(3);

      // Test equivalent to CallTool - single card
      const singleCardResult = await handler.handleToolCall('draw_cards', {});
      expect(singleCardResult.content).toHaveLength(1);

      // Test equivalent to CallTool - multiple cards
      const multipleCardResult = await handler.handleToolCall('draw_cards', { count: 3 });
      expect(multipleCardResult.content).toHaveLength(1);

      // Test equivalent to CallTool - full deck
      const fullDeckResult = await handler.handleToolCall('get_full_deck', {});
      expect(fullDeckResult.content).toHaveLength(1);
    });

    test('handles invalid tool calls correctly', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(42));
      const handler = new TarotMCPToolHandler(deckService);

      const result = await handler.handleToolCall('invalid_tool', {});
      expect(result.isError).toBe(true);
    });

    test('registration logic works correctly', async () => {
      const deckService = new TarotDeckService(new SeedablePseudoRNG(42));
      const handler = new TarotMCPToolHandler(deckService);

      // Test the registration logic that TarotMCPServer.run() uses
      const tools = handler.getAvailableTools();
      expect(tools).toHaveLength(3);

      // Test CallTool registration with various scenarios
      const drawSingleResult = await handler.handleToolCall('draw_cards', {});
      expect(drawSingleResult.content).toHaveLength(1);

      const drawMultipleResult = await handler.handleToolCall('draw_cards', { count: 5 });
      expect(drawMultipleResult.content).toHaveLength(1);
    });
  });
}); 