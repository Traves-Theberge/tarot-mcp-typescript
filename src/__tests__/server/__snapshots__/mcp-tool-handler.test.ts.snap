// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MCP Tool Handler Tests draw_cards tool produces deterministic results for single card 1`] = `
"[
  {
    "name": "Page of Wands",
    "uri": "tarot://card/minor/wands/page"
  }
]"
`;

exports[`MCP Tool Handler Tests get_full_deck tool produces deterministic results 1`] = `
"[
  {
    "name": "The Fool",
    "uri": "tarot://card/major/0"
  },
  {
    "name": "The Magician",
    "uri": "tarot://card/major/1"
  },
  {
    "name": "The High Priestess",
    "uri": "tarot://card/major/2"
  },
  {
    "name": "The Empress",
    "uri": "tarot://card/major/3"
  },
  {
    "name": "The Emperor",
    "uri": "tarot://card/major/4"
  },
  {
    "name": "The Hierophant",
    "uri": "tarot://card/major/5"
  },
  {
    "name": "The Lovers",
    "uri": "tarot://card/major/6"
  },
  {
    "name": "The Chariot",
    "uri": "tarot://card/major/7"
  },
  {
    "name": "Strength",
    "uri": "tarot://card/major/8"
  },
  {
    "name": "The Hermit",
    "uri": "tarot://card/major/9"
  },
  {
    "name": "Wheel of Fortune",
    "uri": "tarot://card/major/10"
  },
  {
    "name": "Justice",
    "uri": "tarot://card/major/11"
  },
  {
    "name": "The Hanged Man",
    "uri": "tarot://card/major/12"
  },
  {
    "name": "Death",
    "uri": "tarot://card/major/13"
  },
  {
    "name": "Temperance",
    "uri": "tarot://card/major/14"
  },
  {
    "name": "The Devil",
    "uri": "tarot://card/major/15"
  },
  {
    "name": "The Tower",
    "uri": "tarot://card/major/16"
  },
  {
    "name": "The Star",
    "uri": "tarot://card/major/17"
  },
  {
    "name": "The Moon",
    "uri": "tarot://card/major/18"
  },
  {
    "name": "The Sun",
    "uri": "tarot://card/major/19"
  },
  {
    "name": "Judgement",
    "uri": "tarot://card/major/20"
  },
  {
    "name": "The World",
    "uri": "tarot://card/major/21"
  },
  {
    "name": "Ace of Wands",
    "uri": "tarot://card/minor/wands/ace"
  },
  {
    "name": "Two of Wands",
    "uri": "tarot://card/minor/wands/2"
  },
  {
    "name": "Three of Wands",
    "uri": "tarot://card/minor/wands/3"
  },
  {
    "name": "Four of Wands",
    "uri": "tarot://card/minor/wands/4"
  },
  {
    "name": "Five of Wands",
    "uri": "tarot://card/minor/wands/5"
  },
  {
    "name": "Six of Wands",
    "uri": "tarot://card/minor/wands/6"
  },
  {
    "name": "Seven of Wands",
    "uri": "tarot://card/minor/wands/7"
  },
  {
    "name": "Eight of Wands",
    "uri": "tarot://card/minor/wands/8"
  },
  {
    "name": "Nine of Wands",
    "uri": "tarot://card/minor/wands/9"
  },
  {
    "name": "Ten of Wands",
    "uri": "tarot://card/minor/wands/10"
  },
  {
    "name": "Page of Wands",
    "uri": "tarot://card/minor/wands/page"
  },
  {
    "name": "Knight of Wands",
    "uri": "tarot://card/minor/wands/knight"
  },
  {
    "name": "Queen of Wands",
    "uri": "tarot://card/minor/wands/queen"
  },
  {
    "name": "King of Wands",
    "uri": "tarot://card/minor/wands/king"
  },
  {
    "name": "Ace of Pentacles",
    "uri": "tarot://card/minor/pentacles/ace"
  },
  {
    "name": "Two of Pentacles",
    "uri": "tarot://card/minor/pentacles/2"
  },
  {
    "name": "Three of Pentacles",
    "uri": "tarot://card/minor/pentacles/3"
  },
  {
    "name": "Four of Pentacles",
    "uri": "tarot://card/minor/pentacles/4"
  },
  {
    "name": "Five of Pentacles",
    "uri": "tarot://card/minor/pentacles/5"
  },
  {
    "name": "Six of Pentacles",
    "uri": "tarot://card/minor/pentacles/6"
  },
  {
    "name": "Seven of Pentacles",
    "uri": "tarot://card/minor/pentacles/7"
  },
  {
    "name": "Eight of Pentacles",
    "uri": "tarot://card/minor/pentacles/8"
  },
  {
    "name": "Nine of Pentacles",
    "uri": "tarot://card/minor/pentacles/9"
  },
  {
    "name": "Ten of Pentacles",
    "uri": "tarot://card/minor/pentacles/10"
  },
  {
    "name": "Page of Pentacles",
    "uri": "tarot://card/minor/pentacles/page"
  },
  {
    "name": "Knight of Pentacles",
    "uri": "tarot://card/minor/pentacles/knight"
  },
  {
    "name": "Queen of Pentacles",
    "uri": "tarot://card/minor/pentacles/queen"
  },
  {
    "name": "King of Pentacles",
    "uri": "tarot://card/minor/pentacles/king"
  },
  {
    "name": "Ace of Swords",
    "uri": "tarot://card/minor/swords/ace"
  },
  {
    "name": "Two of Swords",
    "uri": "tarot://card/minor/swords/2"
  },
  {
    "name": "Three of Swords",
    "uri": "tarot://card/minor/swords/3"
  },
  {
    "name": "Four of Swords",
    "uri": "tarot://card/minor/swords/4"
  },
  {
    "name": "Five of Swords",
    "uri": "tarot://card/minor/swords/5"
  },
  {
    "name": "Six of Swords",
    "uri": "tarot://card/minor/swords/6"
  },
  {
    "name": "Seven of Swords",
    "uri": "tarot://card/minor/swords/7"
  },
  {
    "name": "Eight of Swords",
    "uri": "tarot://card/minor/swords/8"
  },
  {
    "name": "Nine of Swords",
    "uri": "tarot://card/minor/swords/9"
  },
  {
    "name": "Ten of Swords",
    "uri": "tarot://card/minor/swords/10"
  },
  {
    "name": "Page of Swords",
    "uri": "tarot://card/minor/swords/page"
  },
  {
    "name": "Knight of Swords",
    "uri": "tarot://card/minor/swords/knight"
  },
  {
    "name": "Queen of Swords",
    "uri": "tarot://card/minor/swords/queen"
  },
  {
    "name": "King of Swords",
    "uri": "tarot://card/minor/swords/king"
  },
  {
    "name": "Ace of Cups",
    "uri": "tarot://card/minor/cups/ace"
  },
  {
    "name": "Two of Cups",
    "uri": "tarot://card/minor/cups/2"
  },
  {
    "name": "Three of Cups",
    "uri": "tarot://card/minor/cups/3"
  },
  {
    "name": "Four of Cups",
    "uri": "tarot://card/minor/cups/4"
  },
  {
    "name": "Five of Cups",
    "uri": "tarot://card/minor/cups/5"
  },
  {
    "name": "Six of Cups",
    "uri": "tarot://card/minor/cups/6"
  },
  {
    "name": "Seven of Cups",
    "uri": "tarot://card/minor/cups/7"
  },
  {
    "name": "Eight of Cups",
    "uri": "tarot://card/minor/cups/8"
  },
  {
    "name": "Nine of Cups",
    "uri": "tarot://card/minor/cups/9"
  },
  {
    "name": "Ten of Cups",
    "uri": "tarot://card/minor/cups/10"
  },
  {
    "name": "Page of Cups",
    "uri": "tarot://card/minor/cups/page"
  },
  {
    "name": "Knight of Cups",
    "uri": "tarot://card/minor/cups/knight"
  },
  {
    "name": "Queen of Cups",
    "uri": "tarot://card/minor/cups/queen"
  },
  {
    "name": "King of Cups",
    "uri": "tarot://card/minor/cups/king"
  }
]"
`;
